exe/Main.hs

```diff
@@ -188,22 +192,256 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+                -- components in this HscEnv
+                rearrange (uid, (df, uids), ts, cfp, di) = (uid, (df, uids, ts, cfp, di))
+                do_one (uid,df, ts, cfp, di) = rearrange (uid, removeInplacePackages inplace df, ts, cfp, di)
+                -- All deps, but without any packages which are also loaded
+                -- into memory
+                new_deps' = map do_one new_deps
+            -- Make a new HscEnv, we have to recompile everything from
+            -- scratch again (for now)
+            -- It's important to keep the same NameCache though for reasons
+            -- that I do not fully understand
+            print ("Making new HscEnv" ++ (show inplace))
+            hscEnv <- case oldDeps of
+                        Nothing -> emptyHscEnv
+                        Just (old_hsc, _) -> setNameCache (hsc_NC old_hsc) <$> emptyHscEnv
+            newHscEnv <-
+              -- Add the options for the current component to the HscEnv
+              evalGhcEnv hscEnv $ do
+                _ <- setSessionDynFlags df
+                getSession
+            -- Modify the map so the hieYaml now maps to the newly created
+            -- HscEnv
+            -- Returns
+            -- * the new HscEnv so it can be used to modify the
+            --   FilePath -> HscEnv map
+            -- * The information for the new component which caused this cache miss
+            -- * The modified information (without -inplace flags) for
+            --   existing packages
+            pure (Map.insert hieYaml (newHscEnv, new_deps) m, (newHscEnv, head new_deps', tail new_deps'))
+
+
+    session <- return $ \(hieYaml, cfp, opts) -> do
+        (hscEnv, new, old_deps) <- packageSetup (hieYaml, cfp, opts)
+        -- TODO Handle the case where there is no hie.yaml
+        -- Make a map from unit-id to DynFlags, this is used when trying to
+        -- resolve imports.
+        let uids = map (\(iuid, (df, _uis, _targets, _cfp, _di)) -> (iuid, df)) (new : old_deps)
+
+        -- For each component, now make a new HscEnvEq which contains the
+        -- HscEnv for the hie.yaml file but the DynFlags for that component
+        --
+        -- Then look at the targets for each component and create a map
+        -- from FilePath to the HscEnv
+        let new_cache (_iuid, (df, _uis, targets, cfp, di)) =  do
+              let hscEnv' = hscEnv { hsc_dflags = df
+                                   , hsc_IC = (hsc_IC hscEnv) { ic_dflags = df } }
+
+              versionMismatch <- evalGhcEnv hscEnv' checkGhcVersion
+              henv <- case versionMismatch of
+                        Just mismatch -> return mismatch
+                        Nothing -> newHscEnvEq hscEnv' uids
+              let res = (henv, di)
+              print res
+
+              let is = importPaths df
+              ctargets <- concatMapM (targetToFile is  . targetId) targets
+              -- A special target for the file which caused this wonderful
```

Comment:
Why do we need this? Is this to guard against build tools that do not list all targets (e.g. iirc stack doesn’t list the paths modules or something like that?)? Would be great to document this in a comment.

Reply:
Yes but things are very broken anyway if you do not list all the targets. There is potential to improve this in the future by somehow calling downsweep to establish what should implicitly be in the component but it has not been clear to me how to do this yet as the rules currently depend on `GetSession`.